[Tensor refactor] update the Autograd interface to use the new PennyLane tensor class

josh146 opened the pull request:
**Context:**

This is the second PR of the variable refactor, following #648. The previous PR introduced the new `Tensor[ndarray]` subclass, which includes a new `requires_grad` property - allowing NumPy arrays to be marked as differentiable/non-differentiable. By default, all NumPy arrays and Python floats are differentiable.

This follows on by modifying the Autograd interface so that it uses this new property to determine the differentiability of QNode arguments.

**Description of the Change:**

* `qml.grad(func, argnum)` is modified so that the `argnum` argument is now optional. If not provided, the function's non-named arguments are inspected, and any parameters where `requires_grad=False` are dynamically excluded from the gradient computation.

  Note that the `argnum` argument remains, for backwards compatibility with existing code and auxiliary materials that may make use of it.

* The `pennylane.init` module now imports the PennyLane-wrapped NumPy, so that the generated weights are `qml.numpy.tensor` objects. This isn't _strictly_ necessary (since all NumPy arrays, even if not instances of `qml.numpy.tensor`, are assumed to be differentiable), but a convenience for the user so that they have a `requires_grad` property present.

* The autograd interface now checks for non-differentiable positional arguments, and, if present, excludes them from the call to `JacobianQNode.jacobian()`. The VJP is then expanded, with 0s placed wrt the non-differentiable parameters.

* I modified the `JacobianQNode.jacobian` error messages so that the _name_ of the erroneous parameter is returned, rather than the variable index (which is not user-facing information and can be very confusing).

**Benefits:**

* Non-differentiable parameters, when passed as positional arguments, are now _excluded_ from the parameter shift rule.

* More intuitive when working with data embedding and chained QNodes

* No need to manually pass `argnum` to `qml.grad()`, instead it parses the argument list for `requires_grad`.

**Possible Drawbacks:**

* The QNode parameter logic/signature introspection remains, so only positional arguments remain differentiable in the autograd interface. Once the tensor refactor is complete, we can return to the QNode and see what can be deleted. If we are able to make keyword arguments differentiable in `JacobianQNode.jacobian`, we can extend all interfaces likewise.

* `pennylane.numpy` and `pennylane.grad` are now officially parts of the autograd interface. We should consider moving them to the `interfaces` module over time, and make sure they are 'sandboxed' from PennyLane core.

* It feels a bit wrong to have `pennylane.init` import part of the Autograd interface; we have been deliberately moving such that core modules _don't_ depend on Autograd. For Torch and TF, we require users manually cast the output of the `init` module functions to tensors:
  
  ```python
  >>> weights = torch.tensor(strong_ent_layers_normal(n_wires=2, n_layers=3), requires_grad=True)
  >>> weights = tf.Variable(strong_ent_layers_normal(n_wires=2, n_layers=3))
  ```

  should we do the same for the Autograd interface?

  ```python
  >>> weights = np.array(strong_ent_layers_normal(n_wires=2, n_layers=3), requires_grad=True)
  ```
  My gut feeling is no, _but_ we will have to think of something if we ever make Autograd an optional dependence.

**Related GitHub Issues:** #648

co9olguy commented on pennylane/__init__.py:
```diff
@@ -162,23 +162,49 @@ def device(name, *args, **kwargs):
+def grad(func, argnum=None):
     """Returns the gradient as a callable function of (functions of) QNodes.
 
     This is a wrapper around the :mod:`autograd.grad` functions.
```
```suggestion
    This is a wrapper around the :mod:`autograd.grad` function.
```

co9olguy commented on pennylane/__init__.py:
```diff
@@ -162,23 +162,49 @@ def device(name, *args, **kwargs):
+            with respect to. By default, the arguments themselves are used
+            to determine differentiability, by examining the ``requires_grad``
+            property. Providing this keyword argument overrides this behaviour,
+            allowing differentiability to be set manually.
```
It's not really `set` manually, it's just overridden in any one call as specified

josh146 replied:
```suggestion
            allowing argument differentiability to be defined manually for the returned gradient function.
```

co9olguy commented on pennylane/__init__.py:
```diff
@@ -162,23 +162,49 @@ def device(name, *args, **kwargs):
+
+        for idx, arg in enumerate(args):
+            if not getattr(arg, "requires_grad", True):
+                continue
```
is the `continue` necessary here?
```
if getattr(arg, "requires_grad", True):
    argnum.append(idx) 
```

josh146 replied:
Good point 😆

co9olguy commented on pennylane/__init__.py:
```diff
@@ -162,23 +162,49 @@ def device(name, *args, **kwargs):
-            with respect to
+
+    Keyword Args:
+        argnum (int or list(int)): Which argument(s) to take the gradient
```
Add `None` as a type option?

josh146 replied:
```suggestion
        argnum (int or list(int) or None): Which argument(s) to take the gradient
```

co9olguy commented on pennylane/__init__.py:
```diff
@@ -162,23 +162,49 @@ def device(name, *args, **kwargs):
 
     Returns:
         function: the function that returns the gradient of the input
         function with respect to the arguments in argnum
```
might need to update this part of docstring as well?

josh146 replied:
Updated to:
```rest
Returns:
    function: The function that returns the gradient of the input
    function with respect to the differentiable arguments, or, if specified,
    the arguments in ``argnum``.
```

co9olguy commented on pennylane/__init__.py:
```diff
@@ -162,23 +162,49 @@ def device(name, *args, **kwargs):
+                continue
+            argnum.append(idx)
+
+        return _grad(func, argnum)(*args, **kwargs)
```
How come *args and **kwargs need to be specified for this case?

josh146 replied:
The output of `_grad = autograd.grad` is a function that has the same signature as the cost function `func`. In the case of `argnum=None`, we must inspect the `*args` dynamically _within the returned gradient function_, so that we can take into account the case where the user computes the gradient function _once_, but then calls it differently:

```python
# the returned grad_fn is our _gradient_function wrapper
grad_fn = qml.grad(func)
# call grad_fn with both arguments being differentiable
x = np.array([0, 1], requires_grad=True)
y = np.array([0, 1], requires_grad=True)
res1 = grad_fn(x, y)
# call grad_fn with only one argument being differentiable
y.requires_grad = False
res2 = grad_fn(x, y)
```

Essentially, this additional wrapper allows us to move the 'differentiability' check from grad function definition to grad function call time.

co9olguy commented on pennylane/interfaces/autograd.py:
```diff
@@ -67,14 +68,39 @@ def gradient_product(g):
+                diff_indices = None
+                non_diff_indices = set()
+
+                for arg, variable_inds in zip(args, self.arg_vars):
```
is `arg_vars` a poor name choice? It seems that that object actually contains variable indices :thinking:

josh146 replied:
I agree, I was reluctant to change it here as it felt out of scope (it's used in `BaseQNode`, `JacobianQNode`, and `PassthruQNode` I believe)

josh146 replied:
Wait, actually, I was wrong. `self.arg_vars` contains a nested list of `Variable` objects, in the same shape as the input arguments. So if anything, my use of `variable_inds` is incorrect here. I'll rename it to `variables`.

josh146 replied:
Have renamed the loop variable to `arg_variable`, rather than `variable_inds` (which implies an iterable of integers).

co9olguy commented on pennylane/interfaces/autograd.py:
```diff
@@ -67,14 +68,39 @@ def gradient_product(g):
+                        diff_args.append(arg)
+
+                if non_diff_indices:
+                    diff_indices = set(range(self.num_variables)) - non_diff_indices
```
Is this just to avoid the case that a variable that is used in multiple gates might get double counted?

josh146 replied:
Nope, this logic is independent of the underlying QNode operations. Essentially:

1. On construction, the `BaseQNode` takes the input positional arguments `[x[2, 3], y[4], z[1], ...]` (I've included example shapes in square brackets) and flattens it into a single list of `Variables`, with each one corresponding to an _element_ of the input arrays:

   ```pycon
   >>> print(arg_vars)
   [Var(0, name="x[0, 0]"), Var(1, name="x[0, 1]"), ..., Var(9, name="y[3]"), Var(10, name="z[0]")]
   >>> len(_flatten(args))
   >>> 1
   ```

2. It then 'unflattens' this list to match the shape of `args` for convenience, and stores this as `self.arg_vars`:

   ```pycon
   >>> print(self.arg_vars)
   [array([[Var(0, name="x[0, 0]"), Var(1, name="x[0, 1]"), Var(2, name="x[0, 2]")],
           [Var(3, name="x[1, 0]"), Var(4, name="x[1, 1]"), Var(5, name="x[1, 2]")]]),
    array([Var(6, name="y[0]"), Var(7, name="y[1]]"), Var(8, name="y[2]"), Var(9, name="y[3]")]),
    [Var(10, name="z[0]")]]
    ```

All I'm doing here is getting the variable indices of the differentiable parameters, so that I can pass them to `jacobian()` using the `wrt` argument. *Only variables specified in `wrt` will have the parameter shift rule applied*, and the returned Jacobian will be of size `(len(wrt), len(ans))`.

So in this example, if `x.requires_grad = z.requires_grad = False` and `y.requires_grad = True`, then `wrt = {6, 7, 8, 9}`.

josh146 replied:
Internally, if Variable 6 is then used in multiple gates, the `QubitQNode._pd_analytic()` method will internally apply the product rule.

co9olguy commented on pennylane/interfaces/autograd.py:
```diff
@@ -67,14 +68,39 @@ def gradient_product(g):
 
-                # restore the nested structure of the input args
+                if isinstance(args, ndarray):
+                    diff_args = array(diff_args)
```
Won't a `tensor` object satisfy this conditional, and hence the casting to `array` would be redundant?

josh146 replied:
Oops, you're right! I think this is a leftover from a previous iteration of this PR.

josh146 replied:
Have removed these two lines

co9olguy commented on pennylane/interfaces/autograd.py:
```diff
@@ -67,14 +68,39 @@ def gradient_product(g):
+                jac = self.jacobian(args, kwargs, wrt=diff_indices)
+
                 if not g.shape:
                     temp = g * jac  # numpy treats 0d arrays as scalars, hence @ cannot be used
```
let's take the opportunity to give `temp` a better name

josh146 replied:
Have renamed `temp` to `vjp`

co9olguy commented on tests/interfaces/test_autograd.py:
```diff
@@ -475,3 +475,234 @@ def d_classical(a, b, method):
+        """Test that a differentiable parameter used as the first
+        argument is correctly evaluated by QNode.jacobian, and that
+        all other non-differentiable parameters are ignored"""
+        spy = mocker.spy(qml.qnodes.JacobianQNode, "jacobian")
```
:shipit:

co9olguy replied:
Is this a new feature? Haven't seen it used before

josh146 replied:
Kind of, but also not really 😆 

`mocker` is simply a pytest fixture that 'wraps' `unittest.mock`, providing a more convenient UI. So this is doable using `unittest.mock`/subclasses/`with` contexts, but using this fixture makes it significantly more readable I feel.

co9olguy commented on tests/interfaces/test_autograd.py:
```diff
@@ -475,3 +475,234 @@ def d_classical(a, b, method):
+        @qml.qnode(dev, interface="autograd")
+        def circuit(weights, data1, data2):
+            # non-differentiable quantum function
+            qml.templates.AmplitudeEmbedding(data1, wires=[0, 1])
```
Would love to make this one differentiable eventually, but will require some planning for how to implement "in-qnode" classical differentiation

josh146 replied:
Agree :(

co9olguy commented on tests/interfaces/test_autograd.py:
```diff
@@ -475,3 +475,234 @@ def d_classical(a, b, method):
+
+        # check that the parameter shift was only performed for the
+        # differentiable input, not the data input
+        assert spy.call_args[1]["wrt"] == set(data1.size + np.arange(weights.size))
```
Why does `data1.size` enter here? Shouldn't `data1` not appear at all?

josh146 replied:
Ah, I don't think I coded this up in a very clear way. Essentially, I am just checking that
```python
assert spy.call_args[1]["wrt"] == {4, 5, 6, 7, 8, 9}
```
i.e., I'm just using `data1.size` to add on the correct offset (since variables 0-3 should be ignored by the parameter shift rule).

I could have done the offset as
```python
assert spy.call_args[1]["wrt"] == set(np.arange(data1.size, data1.size + weights.size))
```
but I think the 'hardcoded' case above is even better?

co9olguy commented on tests/interfaces/test_autograd.py:
```diff
@@ -475,3 +475,234 @@ def d_classical(a, b, method):
+
+        # check that the parameter shift was only performed for the
+        # differentiable input, not the data input
+        assert spy.call_args[1]["wrt"] == set(data1.size + data2.size + np.arange(weights.size))
```
Same question; the `assert` doesn't seem to intuitively match my expectation from the comment

co9olguy commented on tests/interfaces/test_autograd.py:
```diff
@@ -475,3 +475,234 @@ def d_classical(a, b, method):
+
+        # check that the parameter shift was only performed for the
+        # differentiable input, not the data input
+        diff_variables = spy.call_args[1]["wrt"]
+        expected_diff_variables = set(data1.size + np.arange(weights1.size))
+        expected_diff_variables.update(data1.size + weights1.size + data2.size + np.arange(weights2.size))
+        assert diff_variables == expected_diff_variables
```
I'm finding these `assert`s hard to easily understand

josh146 replied:
I'll rewrite them explicitly :+1:

co9olguy commented on tests/interfaces/test_autograd.py:
```diff
@@ -475,3 +475,234 @@ def d_classical(a, b, method):
+
+        res = grad_fn(weights, data1, data2)
+
+        # we do not check for correctness, just that the output
```
We should maybe check for correctness on some simpler cases?

josh146 replied:
I figure that is already being done by the integration tests above? I would rather this test only fails due to the parameter handling logic/integration failing.

But I will add an additional integration test to check for correctness.

co9olguy commented on tests/interfaces/test_autograd.py:
```diff
@@ -475,3 +475,234 @@ def d_classical(a, b, method):
+        grad_fn = qml.grad(circuit)
+        data1 = qml.numpy.array([0, 1, 1, 0], requires_grad=False) / np.sqrt(2)
+
+        with pytest.warns(UserWarning, match="Output seems independent of input"):
```
Is the warning coming from autograd?
I know it's debated across various libraries, but mathematically, the gradient of f(x) wrt to y is 0. For software/backprop purposes, it may be more convenient to set something else though (`None`, empty tuple, exception)

josh146 replied:
Yep, unfortunately this is coming from Autograd.
https://github.com/HIPS/autograd/blob/65c21e2975fd4d0c6e979c248ee4407e94ed15bf/autograd/tracer.py#L14

co9olguy commented on tests/interfaces/test_autograd.py:
```diff
@@ -475,3 +475,234 @@ def d_classical(a, b, method):
+
+        assert len(res[0]) == 2
+
+
```
```suggestion
```

co9olguy commented on tests/interfaces/test_autograd.py:
```diff
@@ -475,3 +475,234 @@ def d_classical(a, b, method):
+
+        weights = [w1, w2]
+
+        res = cost(weights)
```
```suggestion
```

co9olguy commented on tests/interfaces/test_autograd.py:
```diff
@@ -475,3 +475,234 @@ def d_classical(a, b, method):
+        res = grad_fn(weights)
+
+        assert len(res[0]) == 2
+
```
anything more worth asserting here?

josh146 replied:
Perhaps it's best to check that the gradients are correct, as we don't have any chained QNodes integration tests anywhere.

josh146 commented on pennylane/operation.py:
```diff
@@ -447,6 +447,10 @@ def check_domain(self, p, flattened=False):
         Returns:
             Number, array, Variable: p
         """
+        # If parameter is a NumPy scalar, convert it into a Python scalar.
+        if isinstance(p, np.ndarray) and p.ndim == 0:
+            p = p.item()
```
This change allows NumPy scalars to be used within gate arguments, e.g.
```python
>>> a = np.array(0.5)
>>> circuit(a)
```

PennyLane core already works fine with NumPy scalars, it was just a matter of input validation blocking it. This change is required for the user to mark scalars as non-differentiable when using Autograd:

```python
>>> a = np.array(0.5, requires_grad=False)
>>> circuit(a)
```

and mirrors how PyTorch and TF work (e.g., `tf.Variable(0.5)`).

josh146 commented:
@co9olguy: I've addressed your comments, and updated the PR. In particular:

* I've updated the problematic asserts, let me know if it's clearer or not. The size of the weight arrays was 18, so too big to manually write it by hand?

* I've added two tests; one that checks the gradient is correct when using the new parameter handling system, and another that checks that the gradient of a chained QNode system works correctly.

co9olguy commented on tests/interfaces/test_autograd.py:
```diff
@@ -633,11 +640,15 @@ def circuit(data1, weights1, data2, weights2):
+        offset1 = data1.size
+        offset2 = data1.size + num_w1 + data2.size
+
+        expected = set(list(range(offset1, offset1 + num_w1)) + list(range(offset2, offset2 + num_w2)))
```
since there's no parametrized data here, why not just write explicitly? e.g., `{6,7,8,9,10}`?

josh146 replied:
I was about to, until I realized how long the `wrt` set was 😆 Has over 20 integers in it. Will update 👍

josh146 replied:
done

co9olguy commented on tests/interfaces/test_autograd.py:
```diff
@@ -699,10 +710,115 @@ def cost(weights):
+
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+        # check that the parameter-shift rule was not applied to varphi
```
```suggestion
        # check that the gradient was not applied to varphi
```

co9olguy commented on tests/interfaces/test_autograd.py:
```diff
@@ -699,10 +710,115 @@ def cost(weights):
+            return qml.expval(qml.PauliX(0) @ qml.PauliY(1))
+
+        def cost(a, b, c, weights):
+            return circuit2(circuit1(a, b, c), weights)
```
:heart_eyes:

co9olguy commented on tests/interfaces/test_autograd.py:
```diff
@@ -699,10 +710,115 @@ def cost(weights):
+        # to be compared with the expected result
+        assert np.allclose(np.hstack(res), expected, atol=tol, rtol=0)
+
+        # Check that the parameter-shift rule was applied
+        # to all parameters in circuit2 (i.e., wrt=None)
```
```suggestion
        # Check that the gradient was computed
        # for all parameters in circuit2 (i.e., wrt=None)
```

co9olguy commented on tests/interfaces/test_autograd.py:
```diff
@@ -699,10 +710,115 @@ def cost(weights):
+
+        # Check that the parameter-shift rule was applied
+        # to all parameters in circuit2 (i.e., wrt=None)
+        assert spy.call_args_list[0][1]["wrt"] == None
```
so `"wrt" == None` means wrt all? :smile:

josh146 replied:
Yep! Cases like this are why I have been tempted to make the `AllWires` constant simply `qml.All` that can be used in all relevant places.

Note that this keyword argument has been in PL since the very beginning, but this PR is the very first time that is possible to evaluate a QNode where `wrt != None`

co9olguy commented:
All good to merge @josh146
